File: ruby/hello.rb
require "ed25519"
require 'digest'
require 'sha3'
require "base32"
require 'json'
require "net/http"

signing_key = Ed25519::SigningKey.generate
verify_key = signing_key.verify_key

private_key = signing_key.to_bytes
public_key = verify_key.to_bytes

puts private_key.unpack('H*')
puts public_key.unpack('H*')


#アカウント復元
alice_priavte_key = Ed25519::SigningKey.new(
	"94ee0f4d7fe388ac4b04a6a6ae2ba969617879b83616e4d25710d688a89d80c7" \
	.scan(/../).map{ |b| b.to_i(16) }.pack('C*')
)
alice_public_key = alice_priavte_key.verify_key
puts alice_public_key.to_bytes.unpack('H*')

public_key_hash = SHA3::Digest.hexdigest(:sha256, alice_public_key.to_bytes)
address_body = Digest::RMD160.hexdigest public_key_hash.scan(/../).map{ |b| b.to_i(16) }.pack('C*')
sum_hash = SHA3::Digest.hexdigest(:sha256, ("98" + address_body).scan(/../).map{ |b| b.to_i(16) }.pack('C*'))
alice_address = Base32.encode(("98" + address_body + sum_hash.slice(0..5)).scan(/../).map{ |b| b.to_i(16) }.pack('C*')) 
alice_address = alice_address.slice(0..alice_address.length-2)
puts alice_address


version = [1].pack("C").unpack('H*')[0]
network_type = [152].pack("C").unpack('H*')[0]
transaction_type = [16724].pack("v").unpack('H*')[0]
fee = [1000000].pack("Q").unpack('H*')[0]
second_later_7200 = ((Time.now.to_i + 7200) - 1637848847) * 1000;
deadline = [second_later_7200].pack("Q").unpack('H*')[0]
mosaic_count = [1].pack("C").unpack('H*')[0]
mosaic_id = [0x3A8416DB2D53B6C8].pack("Q").unpack('H*')[0]
recipient_address = Base32.decode("TBIL6D6RURP45YQRWV6Q7YVWIIPLQGLZQFHWFEQ").unpack('H*')[0]
mosaic_amount = [100].pack("Q").unpack('H*')[0]
message = ('Hello Ruby! Welcome to Symbol world!').unpack('H*')[0]
message_size = [message.length / 2 + 1].pack("v").unpack('H*')[0]

verifiable_body = version \
	+ network_type \
	+ transaction_type \
	+ fee \
	+ deadline \
	+ recipient_address \
	+ message_size \
	+ mosaic_count \
	+ "00" + "00000000" \
	+ mosaic_id \
	+ mosaic_amount \
	+ "00" + message


verifiable_string = "7fccd304802016bebbcd342a332f91ff1f3bb5e902988b352697be245f48e836" \
		+ verifiable_body;

verifiable_buffer = verifiable_string.scan(/../).map{ |b| b.to_i(16) }.pack('C*')
signature = alice_priavte_key.sign(verifiable_buffer).unpack('H*')[0]
transaction_size = [verifiable_body.length / 2 + 108].pack("L").unpack('H*')[0]

payload_string = transaction_size \
	+ "00000000" \
	+ signature \
	+ alice_public_key.to_bytes.unpack('H*')[0] \
	+ "00000000" \
	+ verifiable_body

puts payload_string

payload = { "payload" => payload_string}
uri = URI.parse('https://sym-test-02.opening-line.jp:3001/transactions')
req = Net::HTTP::Put.new(uri.request_uri)
req["Content-Type"] = "application/json" 
req.body = payload.to_json

http = Net::HTTP.new(uri.host, uri.port)
http.use_ssl = true
response = http.request(req)
puts response

hashable_string = signature \
	+ alice_public_key.to_bytes.unpack('H*')[0] \
	+ verifiable_string

puts hashable_string
transactionHash  = SHA3::Digest.hexdigest(:sha256, hashable_string.scan(/../).map{ |b| b.to_i(16) }.pack('C*'))

puts "transactionStatus: https://sym-test-02.opening-line.jp:3001/transactionStatus/" + transactionHash;
puts "confirmed: https://sym-test-02.opening-line.jp:3001/transactions/confirmed/" + transactionHash;
puts "explorer: https://testnet.symbol.fyi/transactions/" +  transactionHash;

File: dart/hello.dart
import 'dart:convert';
import 'dart:developer';
import 'dart:typed_data';
import 'package:sha3/sha3.dart';
import 'package:base32/base32.dart';
import 'package:convert/convert.dart';
import 'package:http/http.dart' as http;
import 'package:ed25519_edwards/ed25519_edwards.dart' as ed;
import 'package:hash/hash.dart';

void main() {

	var keyPair = ed.generateKey();
	var privateKey = keyPair.privateKey;
	var publicKey = keyPair.publicKey;
	print(hex.encode(privateKey.bytes));
	print(hex.encode(publicKey.bytes));

	var alicePrivateKey = ed.newKeyFromSeed(hex.decode("94ee0f4d7fe388ac4b04a6a6ae2ba969617879b83616e4d25710d688a89d80c7"));
	print(alicePrivateKey.bytes);

	var alicePublicKey = ed.public(alicePrivateKey);
	var aliceKeypair = ed.KeyPair(alicePrivateKey,alicePublicKey);

	print(hex.encode(alicePrivateKey.bytes));
	print(hex.encode(alicePublicKey.bytes));

	var addressHasher = SHA3(256, SHA3_PADDING, 256);
	var publicKeyHash = addressHasher.update(alicePublicKey.bytes).digest();
	print("publicKeyHash");
	print(hex.encode(publicKeyHash));

	var addressBody = RIPEMD160().update(publicKeyHash).digest();
	print("addressBody");
	print(hex.encode(addressBody));

	var sumHasher = SHA3(256, SHA3_PADDING, 256);
	var sumHash = sumHasher.update(hex.decode("98" + hex.encode(addressBody))).digest();
	print(hex.encode(sumHash).substring(0,6));
	var aliceAddress = base32.encode(hex.decode("98" + hex.encode(addressBody) + hex.encode(sumHash).substring(0,6)));
	print("98" + hex.encode(addressBody) + hex.encode(sumHash).substring(0,6));
	aliceAddress = aliceAddress.substring(0, aliceAddress.length - 1);
	print(aliceAddress);

	var version = Uint8List(1)..buffer.asByteData().setInt8(0, 1);
	var networkType  = Uint8List(1)..buffer.asByteData().setInt8(0, 152);
	var transactionType   = Uint8List(2)..buffer.asByteData().setInt16(0, 16724, Endian.little);
	var fee   = Uint8List(8)..buffer.asByteData().setInt64(0, 1000000, Endian.little);
	var dt = (new DateTime.now().millisecondsSinceEpoch / 1000).floor();
	var secondLater7200 = ((dt + 7200) - 1637848847) * 1000;
	var deadline   = Uint8List(8)..buffer.asByteData().setInt64(0, secondLater7200, Endian.little);
	var recipientAddress = base32.decodeAsHexString("TBIL6D6RURP45YQRWV6Q7YVWIIPLQGLZQFHWFEQ");
	print(recipientAddress);
	var mosaicCount = Uint8List(1)..buffer.asByteData().setInt8(0, 1);
	var mosaicId = Uint8List(8)..buffer.asByteData().setInt64(0, 0x3A8416DB2D53B6C8, Endian.little);
	var mosaicAmount = Uint8List(8)..buffer.asByteData().setInt64(0, 100, Endian.little);
	List<int> message = utf8.encode('Hello Dart! Welcome to Symbol world!');
	var messageSize    = Uint8List(2)..buffer.asByteData().setInt16(0, message.length + 1, Endian.little);

	var verifiableBody = hex.encode(version)
		+ hex.encode(networkType)
		+ hex.encode(transactionType)
		+ hex.encode(fee)
		+ hex.encode(deadline)
		+ recipientAddress
		+ hex.encode(messageSize)
		+ hex.encode(mosaicCount)
		+ "00" + "00000000"
		+ hex.encode(mosaicId)
		+ hex.encode(mosaicAmount)
		+ "00" + hex.encode(message);

	var verifiableString = "7fccd304802016bebbcd342a332f91ff1f3bb5e902988b352697be245f48e836"
		+ verifiableBody;

	var verifiableBuffer = hex.decode(verifiableString);
	var signature = ed.sign(alicePrivateKey, verifiableBuffer);

	var transactionSize = Uint8List(4)..buffer.asByteData().setInt32(0, hex.decode(verifiableBody).length + 108, Endian.little);

	var payloadString = hex.encode(transactionSize)
		+ "00000000"
		+ hex.encode(signature) 
		+ hex.encode(alicePublicKey.bytes) 
		+ "00000000" 
		+ verifiableBody;

	var payload = {
		"payload":payloadString
	};
/*
	var response = http.put(
		Uri.parse("https://sym-test-02.opening-line.jp:3001/transactions"),
		headers: {'Content-Type': 'application/json'},
		body: jsonEncode(payload),
	)
	.then((http.Response response) {
		print("Response status: ${response.statusCode}");
		print("Response body: ${response.contentLength}");
		print(response.headers);
		print(response.request);
	});
*/

	var hashableBuffer = hex.decode(
		 hex.encode(signature) 
		+ hex.encode(alicePublicKey.bytes) 
		+ verifiableString
	);

	var hasher = SHA3(256, SHA3_PADDING, 256);
	var transactionHash = hasher.update(hashableBuffer).digest();

	print("transactionStatus: https://sym-test-02.opening-line.jp:3001/transactionStatus/" + hex.encode(transactionHash));
	print("confirmed: https://sym-test-02.opening-line.jp:3001/transactions/confirmed/" + hex.encode(transactionHash));
	print("explorer: https://testnet.symbol.fyi/transactions/" +  hex.encode(transactionHash));
}
